peas/weather_abstract.py

```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env python3
+
+import logging
+import numpy as np
+import re
+import serial
+import sys
+import time
+import requests
+
+from datetime import datetime as dt
+from dateutil.parser import parse as date_parser
+
+import astropy.units as u
+from astropy.units import cds
+from astropy.table import Table
+from astropy.time import Time, TimeISO, TimeDelta
+
+from pocs.utils.messaging import PanMessaging
+
+from . import load_config
+from .PID import PID
+
+def get_mongodb():
+    from pocs.utils.database import PanMongo
+    return PanMongo()
+
+# -----------------------------------------------------------------------------
+#   Base class to read and check weather data
+# -----------------------------------------------------------------------------
+class WeatherAbstract(object):
+
+    """ Base class for checking generic weather data and sending it to the
+        required location.
+    """
+
+    def __init__(self, use_mongo=True):
+        self.config = load_config()
+
+        self.thresholds = self.config.get('thresholds')
+        self.safety_delay = self.config.get('safety_delay', 15.)
+
+        self.db = None
+        if use_mongo:
+            self.db = get_mongodb()
+
+        self.messaging = None
+        self.safe_dict = None
+        self.weather_entries = list()
+
+    def send_message(self, msg, channel='weather'):
+        if self.messaging is None:
+            self.messaging = PanMessaging.create_publisher(6510)
+
+        self.messaging.send_message(channel, msg)
+
+    def capture(self, data):
+        # Make Safety Decision
+        self.safe_dict = self.make_safety_decision(data)
+
+        data['Safe'] = self.safe_dict['Safe']
+        data['Sky condition'] = self.safe_dict['Sky']
+        data['Wind condition'] = self.safe_dict['Wind']
+        data['Gust condition'] = self.safe_dict['Gust']
+        data['Rain condition'] = self.safe_dict['Rain']
+
+        # Store current weather
+        self.weather_entries.append(data)
+
+        if send_message:
+            self.send_message({'data': data}, channel='weather')
+
+        if use_mongo:
+            self.db.insert_current('weather', data)
+
+        return data
+
+    def make_safety_decision(self, current_values):
+        self.logger.debug('Making safety decision')
+        self.logger.debug('Found {} weather data entries in last {:.0f} minutes'.format(
+            len(self.weather_entries), self.safety_delay))
+        safe = False
+
+        # Tuple with condition,safety
+        cloud = self._get_cloud_safety(current_values)
+
+        try:
+            wind, gust = self._get_wind_safety(current_values)
+        except Exception as e:
+            self.logger.warning('Problem getting wind safety: {}'.format(e))
+            wind = ['N/A']
+            gust = ['N/A']
+
+        rain = self._get_rain_safety(current_values)

```

Comment:
The abstract base class doesn't (and shouldn't) have a `_get_rain_safety()` method so this will fail with an `AttributeError` in the base class or in any derived class that doesn't implement `_get_rain_safety()`.

If you think that `_get_rain_safety()` should be implemented in all derived classes (i.e. all weather stations will always have a rain sensor of some kind) then the correct way to handle this is to put a `_get_rain_safety()` method in the abstract base class that contains only `raise NotImplementedError`.  That makes it clear to someone writing a derived class that they need to write a `_get_rain_safety()` method, and ensures that there is a reasonably informative error message if they don't.